src/core/lib/event_engine/forkable.cc

```diff
@@ -43,6 +44,9 @@ Forkable::~Forkable() { StopManagingForkable(this); }
 
 void RegisterForkHandlers() {
   grpc_core::MutexLock lock(g_mu.get());
+#ifdef GRPC_ONLY_REGISTER_PTHREAD_ATFORK_WITH_ENV_VAR
+  if (!grpc_core::ConfigVars::Get().EnableForkSupport()) return;
```

Comment:
> Why is runtime configuration necessary here? Mainly because you want to produce a single binary that can run in both modes?

Yes we do ship pre-compiled binaries that need to work with and without fork support.

But essentially, enabling fork support in gRPC/Ruby constitutes a breaking change. So it needs to use some explicit opt-in (I'm using the environment variable for that here).

Reason it's a breaking change is because, today, before this PR, people can spawn child processes from gRPC/Ruby processes without issue (i.e. fork-with-immediate-exec poses no problems with gRPC/Ruby today).

But as soon as we register C-core's `pthread_atfork` handlers, we can no longer spawn such child processes without calling `GRPC.prefork` before and `GRPC.postfork_parent` afterwards (the ruby API: `GRPC.prefork` is needed to shutdown gRPC/Ruby background threads, which would otherwise prevent C-core's `pthread_atfork` handlers from working properly, because there's otherwise a possibility that the ruby background threads could still be calling into C-core at the time of fork).

Correct me if I'm wrong @gnossen, but my understanding from our chat so far was that python needs the `pthread_atfork` handlers to always be registered, because python users can expect to be able to fork-without-exec *without specifying an env var*, i.e. if we guarded this by an env var then we could break python users.

Reply:
Just thinking about this, another possibility is that we could always register the `pthread_atfork` handlers in C-core, but maybe we could have some global variable, e.g. `g_enable_pthread_atfork`, that could determine whether or not C-core's `pthread_atfork` handlers do anything (e.g. they will do nothing if `g_enable_pthread_atfork == false`).

Then there could be some C-core API, e.g. `grpc_enable_pthread_atfork_handlers` and `grpc_enable_pthread_atfork_handlers`, to toggle `g_enable_pthread_atfork`.

So Ruby's `GRPC.prefork` API could call `grpc_enable_pthread_atfork_handlers`, and we could disable them again in `GRPC.postfork`.

I bring this up because that might give us a way to enable fork support without env vars in ruby.

Reply:
To give a user perspective, the env var isn't a deal breaker, but it's would really help usability if it wasn't strictly required.

Reply:
I think reintroducing an env-var in Forkable is probably the right thing here, presuming both wrapped languages expect most fork operations to be fork+exec. The gRPC Python community already expects to have to use the env var, so it's not a big change. However, I'm not sure the Python build is doing this correctly today. See analysis below.

@apolcyn I'm not sure how deeply Ruby can reach into core, but `ConfigVars::SetOverrides` exists if you want to override fork support programatically.

----

IIUC, non-windows Python builds enable fork support by default in core, so the env var is redundant _in core_ unless you're using it to turn fork support off. https://github.com/grpc/grpc/blob/1bf3f56c9798e8e162ffe1383bcef22580e1defb/setup.py#L459-L462

pthread support is then enabled for all linux & darwin systems
https://github.com/grpc/grpc/blob/1bf3f56c9798e8e162ffe1383bcef22580e1defb/setup.py#L479-L485

gRPC Python code, by default, does not engage its own fork handlers unless the env var is present. https://github.com/grpc/grpc/blob/1bf3f56c9798e8e162ffe1383bcef22580e1defb/src/python/grpcio/grpc/_cython/_cygrpc/fork_posix.pyx.pxi#L32-L34

@gnossen the question then becomes: what are the scenarios in which you want to disable fork support in Python, while core is still using pthread fork handlers? I believe that is what's happening by default in both iomgr & EventEngine fork support today.